Updated Server Logger and OnSession functions (#48)
tmaxmax/go-sse: CHANGELOG.md

This file tracks changes to this project. It follows the [Keep a Changelog format](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] - 2025-02-02

> Logging and Server request handling revamped to hopefully be more familiar interfaces for users.

Reviewer: ```suggestion
The `sse.Server` logging and session handling were revamped to have more familiar, more flexible and less error prone interfaces for users.
```

### Removed

> - Logger and LogLevel enum have been removed.

Reviewer: Just some backticks:
```suggestion
- `Logger` and `LogLevel` enum have been removed. `Server.Logger` has transitioned to the standard `slog` library for better compatibility with the ecosystem
```

### Changed

> - Server.Logger is now of type `func(r *http.Request) *slog.Logger` instead of `sse.Logger`

Reviewer: Backticks and some more details:
```suggestion
- `Server.Logger` is now of type `func(r *http.Request) *slog.Logger` instead of `sse.Logger` – it is possible to customize the logger on a per-request basis, by for example retrieving it from the context. 
```

> - Server.OnSession signature changed from `func(s *sse.Session) (Subscription, bool)` to `func(w http.ResponseWriter, r
> *http.Request) (topics []string, accepted bool)`

Reviewer: Backticks and more detail
```suggestion
- `Server.OnSession` signature changed from `func(s *Session) (Subscription, bool)` to `func(w http.ResponseWriter, r *http.Request) (topics []string, accepted bool)` – its initial role was to essentially just provide the topics, so the need to fiddle with `Session` and `Subscription` was redundant anyway
```

> - complex example updated to demonstrate new behavior

Reviewer: That's technically not a change to the library itself so let's not cover it in the changelog.
```suggestion
```

- `Server.Logger` is now of a new type: the `Logger` interface. The dependency on x/exp/slog is removed. This opens up the possibility to adapt any existing logger to be usable with `Server`.
- The default backoff behavior has changed. The _previous_ defaults map to the new `Backoff` configuration as follows:

> 

Reviewer: Can we remove these formatting changes or include them in a separate commit?

Author: ✅ removed formatting changes
